### ISSUE TITLE
Connection Reset by Peer causes Frenetic Crash

### ISSUE THREAD

**Issue by craig-riecke** [frenetic-lang/frenetic#422]
This may be related to Issue #374, but the stack trace seems different.  It's unclear what network event caused this to happen.  128.84.155.54 34716 is probably the interesting IP address/port here.

```
((pid 21089) (thread_id 0)
 ((human_readable 2015-06-15T14:17:20+0000)
  (int63_ns_since_epoch 1434392240576921000))
 "unhandled exception in Async scheduler"
 ("unhandled exception"
  ((lib/monitor.ml.Error_
    ((exn
      ("writer error"
       ((lib/monitor.ml.Error_
         ((exn
           (Unix.Unix_error "Connection reset by peer"
            writev_assume_fd_is_nonblocking ""))
          (backtrace
           ("Raised at file \"lib/writer0.ml\", line 582, characters 49-52"
            "Called from file \"lib/job_queue.ml\", line 164, characters 6-47"
            ""))
          (monitor
           (((name (id 84877)) (here ()) (id 84877) (has_seen_error true)
             (is_detached true) (kill_index 0))
            ((name main) (here ()) (id 1) (has_seen_error true)
             (is_detached false) (kill_index 0))))))
        ((id 1878)
         (fd
          ((file_descr 36)
           (info
            (socket
             ((listening_on
               ((type_
                 ((family
                   ((family PF_INET) (address_of_sockaddr_exn <fun>)
                    (sexp_of_address <fun>)))
                  (socket_type SOCK_STREAM)))
                (fd
                 ((file_descr 8)
                  (info
                   (replaced
                    (listening
                     (previously_was
                      (replaced
                       ((socket (bound_on (Inet (0.0.0.0 6633))))
                        (previously_was
                         (socket
                          ((family
                            ((family PF_INET) (address_of_sockaddr_exn <fun>)
                             (sexp_of_address <fun>)))
                           (socket_type SOCK_STREAM))))))))))
                  (kind (Socket Passive)) (supports_nonblock true)
                  (have_set_nonblock true) (state (Close_requested <fun>))
                  (watching ((read Stop_requested) (write Not_watching)))
                  (watching_has_changed true) (num_active_syscalls 1)
                  (close_finished Empty)))))
              (client (Inet (128.84.155.54 34716))))))
           (kind (Socket Active)) (supports_nonblock true)
           (have_set_nonblock true) (state Closed)
           (watching ((read Not_watching) (write Not_watching)))
           (watching_has_changed false) (num_active_syscalls 0)
           (close_finished (Full ()))))
         (monitor
          (((name (id 84876)) (here ()) (id 84876) (has_seen_error true)
            (is_detached true) (kill_index 0))
           ((name main) (here ()) (id 1) (has_seen_error true)
            (is_detached false) (kill_index 0))))
         (inner_monitor
          (((name (id 84877)) (here ()) (id 84877) (has_seen_error true)
            (is_detached true) (kill_index 0))
           ((name main) (here ()) (id 1) (has_seen_error true)
            (is_detached false) (kill_index 0))))
         (background_writer_state Stopped_permanently) (syscall Per_cycle)
         (bytes_received 16) (bytes_written 8) (scheduled <opaque>)
         (scheduled_bytes 0) (buf <opaque>) (scheduled_back 0) (back 0)
         (flushes <opaque>) (close_state Closed) (close_finished (Full ()))
         (producers_to_flush_at_close ()) (flush_at_shutdown_elt (<opaque>))
         (check_buffer_age
          (((writer <opaque>) (queue ()) (maximum_age 2m) (too_old Empty))))
         (consumer_left (Full ())) (raise_when_consumer_leaves true)
         (open_flags (Full (Ok (rdwr))))))))
     (monitor
      (((name (id 84876)) (here ()) (id 84876) (has_seen_error true)
        (is_detached true) (kill_index 0))
       ((name main) (here ()) (id 1) (has_seen_error true)
        (is_detached false) (kill_index 0))))))
   (Pid 21089))))
'''

**Comment by areteix** [comment]
That's the wireless router trying to connect to the controller.
```
Sep 04 20:53:17|00010|port_watcher|INFO|Datapath id is 000000000001
Sep 04 20:53:47|00011|rconn|ERR|tcp:10.22.244.8:6633: no response to inactivity probe after 15 seconds, disconnecting
Sep 04 20:53:47|00012|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:53:48|00013|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:53:49|00014|rconn|INFO|tcp:10.22.244.8:6633: connected
Sep 04 20:54:20|00015|rconn|ERR|tcp:10.22.244.8:6633: no response to inactivity probe after 15 seconds, disconnecting
Sep 04 20:54:20|00016|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:54:21|00017|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:54:22|00018|rconn|INFO|tcp:10.22.244.8:6633: connected
Sep 04 20:54:52|00019|rconn|ERR|tcp:10.22.244.8:6633: no response to inactivity probe after 15 seconds, disconnecting
Sep 04 20:54:52|00020|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:54:53|00021|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:54:54|00022|rconn|INFO|tcp:10.22.244.8:6633: connected
Sep 04 20:55:24|00023|rconn|ERR|tcp:10.22.244.8:6633: no response to inactivity probe after 15 seconds, disconnecting
Sep 04 20:55:24|00024|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:55:25|00025|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:55:26|00026|rconn|INFO|tcp:10.22.244.8:6633: connection timed out
Sep 04 20:55:26|00027|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:55:26|00028|rconn|INFO|tcp:10.22.244.8:6633: waiting 2 seconds before reconnect
Sep 04 20:55:28|00029|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:55:30|00030|rconn|INFO|tcp:10.22.244.8:6633: connection timed out
Sep 04 20:55:30|00031|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:55:30|00032|rconn|INFO|tcp:10.22.244.8:6633: waiting 4 seconds before reconnect
Sep 04 20:55:34|00033|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:55:38|00034|rconn|INFO|tcp:10.22.244.8:6633: connection timed out
Sep 04 20:55:38|00035|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:55:38|00036|rconn|INFO|tcp:10.22.244.8:6633: waiting 8 seconds before reconnect
Sep 04 20:55:46|00037|rconn|INFO|tcp:10.22.244.8:6633: connecting...
Sep 04 20:55:54|00038|rconn|INFO|tcp:10.22.244.8:6633: connection timed out
Sep 04 20:55:54|00039|rconn|INFO|tcp:10.22.244.8:6633: connection dropped
Sep 04 20:55:54|00040|rconn|INFO|tcp:10.22.244.8:6633: waiting 15 seconds before reconnect
Sep 04 20:55:55|00041|rconn|INFO|tcp:127.0.0.1:6634: connection closed by peer
Sep 04 20:55:56|00042|rconn|INFO|tcp:127.0.0.1:6634: connecting...
Sep 04 20:55:56|00043|rconn|INFO|tcp:127.0.0.1:6634: connected

```

**Comment by craig-riecke** [comment]
This was actually solved by #416.  (1) There are no more Connection Reset by Peer messages in the log.  (2) I concocted a Python script that emulates an OpenFlow switch, then disconnects with a hard RST (which is the way you make a Connection Reset by Peer happen).  The Openflow process ran without exception, and when I reconnected, it did just fine.

**Comment by craig-riecke** [comment]
Here is the Python program I used to test, for posterity:

```python
# Echo client program
import socket

from ryu.ofproto.ofproto_v1_0_parser import *
from ryu.ofproto.ofproto_v1_0 import *
from ryu.ofproto import ofproto_v1_0, ofproto_v1_0_parser

class Datapath(object):
  ofproto = ofproto_v1_0
  ofproto_parser = ofproto_v1_0_parser

class OFPSwitchFeaturesSerializable(OFPSwitchFeatures):
  def _serialize_body(self):
    msg_pack_into(OFP_SWITCH_FEATURES_PACK_STR, self.buf,
      OFP_HEADER_SIZE, self.datapath_id, self.n_buffers, self.n_tables,
      self.capabilities, self.actions)

HOST = 'localhost'    # The remote host
PORT = 6633              # The same port as used by the server
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((HOST, PORT))

# This is so we can send a RESET connection
l_onoff = 1                                                                                                                                                           
l_linger = 0                                                                                                                                                          
s.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER,                                                                                                                     
             struct.pack('ii', l_onoff, l_linger))

data = s.recv(8)
print 'Received', repr(data)

# This should be the OFPT_HELLO message.  So we send one right back
s.sendall('\x01\x00\x00\x08\x00\x00\x00\x00')
data = s.recv(8)
print 'Received', repr(data)

# At this point, we should have received a request for Switch Features, so we send a minimal one
openflow_constants = dir(ofproto_v1_0)

msg = OFPSwitchFeaturesSerializable(Datapath,
  datapath_id = 9210263729383,
  n_buffers = 897345987,
  n_tables = 250,
  capabilities = OFPC_FLOW_STATS | OFPC_TABLE_STATS,
  actions = OFPAT_OUTPUT
)
msg.serialize()
s.sendall(msg.buf)
data = s.recv(1024)
print 'Received', repr(data)

# This should cause a Connection Reset By Peer error
s.close()
```
